controllers/cluster_controller.go

```diff
@@ -278,70 +296,219 @@ func (r *ClusterReconciler) upgradeControlPlane(ctx context.Context, cluster *ca
 	return controlPlaneNodesWillBeRolled, nil
 }
 
-func (r *ClusterReconciler) upgradeWorkers(ctx context.Context, cluster *capi.Cluster, giantswarmRelease *releaseapiextensions.Release) error {
+// upgradeMachinePool checks the machine image and k8s version used and will
+// update the infra Machine Pool so it uses the right versions.
+func (r *ClusterReconciler) upgradeMachinePool(ctx context.Context, cluster *capi.Cluster, machinePool *capiexp.MachinePool, giantswarmRelease *releaseapiextensions.Release) (bool, error) {
+	logger := r.Log.WithValues("cluster", cluster.Name, "machinepool", machinePool.Name)
+	logger.Info("Fetching infrastructure MachinePool to check its k8s version and machine image")
+
 	expectedCAPIVersion := getComponentVersion(giantswarmRelease, capiReleaseComponent)
 	expectedK8sVersion := getComponentVersion(giantswarmRelease, "kubernetes")
 	expectedOSVersion := getComponentVersion(giantswarmRelease, "image")
 	expectedMachineImage := MachineImagesByK8sVersion[expectedK8sVersion][expectedOSVersion]
 
-	clusterMachinePools, err := r.getSortedClusterMachinePools(ctx, cluster)
+	infraMachinePool, err := external.Get(ctx, r.Client, &machinePool.Spec.Template.Spec.InfrastructureRef, machinePool.Namespace)
 	if err != nil {
-		return err
+		return false, err
 	}
 
-	for _, machinePool := range clusterMachinePools {
-		logger := r.Log.WithValues("cluster", cluster.Name, "machinepool", machinePool.Name)
-		logger.Info("Fetching infrastructure MachinePool to check its k8s version and machine image")
+	currentMachineImage, _, err := unstructured.NestedString(infraMachinePool.Object, strings.Split(KindToMachineImagePath[infraMachinePool.GetKind()], ".")...)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to read current machine image from infrastructure machine pool template %q", infraMachinePool.GetName())
+	}
 
-		infraMachinePool, err := external.Get(ctx, r.Client, &machinePool.Spec.Template.Spec.InfrastructureRef, machinePool.Namespace)
+	// If any MachinePool is not ready we exit because we don't want to
+	// trigger an upgrade while something is wrong or another upgrade is in
+	// progress.
+	if !isReady(machinePool) {
```

Comment:
Minor nit: this could be done earlier before `getComponentVersion` calls as if we exit early those calls or accessing infra machine pool are not necessary.

Reply:
True, I'll do the check earlier

Reply:
Done